SignalServiceKit/src/Messages/Interactions/TSQuotedMessage.m

```diff
@@ -3,46 +3,329 @@
+                                   TSIncomingMessage *incomingMessage = (TSIncomingMessage *)interaction;
+                                   return [authorId isEqual:incomingMessage.messageAuthorId];
+                               } else if ([interaction isKindOfClass:[TSOutgoingMessage class]]) {
+                                   return [authorId isEqual:[TSAccountManager localNumber]];
+                               } else {
+                                   // ignore other interaction types
+                                   return NO;
+                               }
+
+                           }
+                  withTransaction:transaction];
+
+    TSMessage *_Nullable quotedMessage = quotedMessages.firstObject;
+
+    if (!quotedMessage) {
+        return nil;
+    }
+
+    TSAttachment *attachment = [quotedMessage attachmentWithTransaction:transaction];
+    if (![attachment isKindOfClass:[TSAttachmentStream class]]) {
+        return nil;
+    }
+    TSAttachmentStream *sourceStream = (TSAttachmentStream *)attachment;
+
+    TSAttachmentStream *_Nullable thumbnailStream = [sourceStream cloneAsThumbnail];
+    if (!thumbnailStream) {
+        return nil;
+    }
+
+    return thumbnailStream;
+}
+
+
+#pragma mark - Attachment (not necessarily with a thumbnail)
+
+- (nullable OWSAttachmentInfo *)firstAttachmentInfo
+{
+    OWSAssert(self.quotedAttachments.count <= 1);
+    return self.quotedAttachments.firstObject;
+}
+
+- (nullable NSString *)contentType
+{
+    OWSAttachmentInfo *firstAttachment = self.firstAttachmentInfo;
+
+    return firstAttachment.contentType;
+}
+
+- (nullable NSString *)sourceFilename
+{
+    OWSAttachmentInfo *firstAttachment = self.firstAttachmentInfo;
+
+    return firstAttachment.sourceFilename;
+}
+
+- (nullable NSString *)thumbnailAttachmentPointerId
+{
+    OWSAttachmentInfo *firstAttachment = self.firstAttachmentInfo;
+
+    return firstAttachment.thumbnailAttachmentPointerId;
+}
+
+- (nullable NSString *)thumbnailAttachmentStreamId
+{
+    OWSAttachmentInfo *firstAttachment = self.firstAttachmentInfo;
+
+    return firstAttachment.thumbnailAttachmentStreamId;
 }
 
+- (void)setThumbnailAttachmentStream:(TSAttachmentStream *)attachmentStream
+{
+    OWSAssert([attachmentStream isKindOfClass:[TSAttachmentStream class]]);
+    OWSAssert(self.quotedAttachments.count == 1);
+
+    OWSAttachmentInfo *firstAttachment = self.firstAttachmentInfo;
+    firstAttachment.thumbnailAttachmentStreamId = attachmentStream.uniqueId;
+}
+
+- (NSArray<NSString *> *)thumbnailAttachmentStreamIds
+{
+    NSMutableArray *streamIds = [NSMutableArray new];
+    for (OWSAttachmentInfo *info in self.quotedAttachments) {
+        if (info.thumbnailAttachmentStreamId) {
+            [streamIds addObject:info.thumbnailAttachmentStreamId];
+        }
+    }
+
+    return [streamIds copy];
+}
+
+- (NSArray<TSAttachmentStream *> *)createThumbnailAttachmentsIfNecessaryWithTransaction:
```

Comment:
Where is this method called?

Reply:
In OWSMesageSender, before sending. If we persist the thumbnails before that, they risk getting orphaned.

I'll add a comment.